File: src/LoanViewer.Api/Services/LoanCalculationService.cs
using LoanViewer.Api.Models;

namespace LoanViewer.Api.Services;

public class LoanCalculationService
{
    public double Amount { get; private set; }
    public double Rate { get; private set; }
    public int Terms { get; private set; }

    public AmortizationTable AmortizationTable => GetAmortizationTable();
    public string? MonthlyPayment => "$" + GetMonthlyPayment().ToString("N");
    public string? YearlyPayment => "$" + GetYearlyPayment().ToString("N");
    public string? TotalInterest =>
        "$" + AmortizationTable.AmortizationTableItems.Select(x => x.InterestPaid).Sum().ToString("N");

    public void SetLoanDetails(
        double amount,
        double rate,
        int terms,
        LoanTerms.TermType term = LoanTerms.TermType.Yearly
    )
    {
        var loanTermns = new LoanTerms
        {
            Amount = amount,
            Rate = rate,
            Duration = terms,
            Term = term
        };
        SetLoanDetails(loanTermns);
    }

    public void SetLoanDetails(LoanTerms loanTerms)
    {
        var amount = loanTerms.Amount;
        var rate = loanTerms.Rate;
        var terms = loanTerms.Duration;

        Amount = amount;
        Terms = loanTerms.Term == LoanTerms.TermType.Yearly ? terms : terms / 12;
        Rate = rate < 1 ? rate : rate / 100;
    }

    private double GetYearlyPayment() =>
        Amount * ((Rate * Math.Pow(1 + Rate, Terms)) / (Math.Pow(1 + Rate, Terms) - 1));

    private double GetMonthlyPayment() => GetYearlyPayment() / 12;

    private AmortizationTable GetAmortizationTable()
    {
        var amortizationTable = new AmortizationTable
        {
            MonthlyPayment = GetMonthlyPayment(),
            YearlyPayment = GetYearlyPayment()
        };
        var beginningBalance = Amount;
        var totalInterestPaid = 0.00;
        var yearlyPayment = GetYearlyPayment();
        for (int i = 0; i < Terms; i++)
        {
            var interestPaid = beginningBalance * Rate;
            totalInterestPaid += interestPaid;
            var principalPaid = yearlyPayment - interestPaid;
            var remainingPrincipal = beginningBalance - principalPaid;
            remainingPrincipal = (remainingPrincipal < 1) ? 0 : remainingPrincipal;
            amortizationTable.AmortizationTableItems.Add(
                new AmortizationTableItem(
                    Term: i + 1,
                    PrincipalPaid: principalPaid,
                    InterestPaid: interestPaid,
                    RemainingPrincipal: remainingPrincipal,
                    BeginningBalance: beginningBalance,
                    EndingBalance: remainingPrincipal
                )
            );
            beginningBalance = remainingPrincipal;
        }
        return amortizationTable;
    }
}

File: src/LoanViewer.Web/src/services/LoanCalculationService.ts
import {
  AmountType,
  TermType,
  type LoanFormModel,
  type LoanModel,
} from "@/Models/LoanModel";

export class LoanCalculationService {
  amount: number;
  rate: number;
  terms: number;

  constructor(model: LoanFormModel) {
    if (model.interestRate > 1) {
      model.interestRate /= 100;
    }

    this.amount = model.amount;
    this.rate = model.interestRate;
    this.terms = model.term;

    if (
      model.options.amountType === AmountType.Monthly &&
      model.options.termType === TermType.Months
    ) {
      this.amount = model.amount * this.terms;
      this.rate /= 12;
    } else if (
      model.options.amountType === AmountType.Monthly &&
      model.options.termType === TermType.Years
    ) {
      this.amount = model.amount * this.terms * 12;
    } else if (
      model.options.amountType === AmountType.Full &&
      model.options.termType === TermType.Months
    ) {
      this.rate /= 12;
    }

    console.log("Loan Service", model);
  }

  get amortizationTable(): LoanModel {
    return this.getAmortizationTable();
  }

  public getMonthlyPayment(): number {
    return this.getYearlyPayment() / 12;
  }

  public getYearlyPayment(): number {
    return (
      this.amount *
      ((this.rate * Math.pow(1 + this.rate, this.terms)) /
        (Math.pow(1 + this.rate, this.terms) - 1))
    );
  }

  public getAmortizationTable(): LoanModel {
    const amortizationTable: LoanModel = {
      monthlyPayment: this.getMonthlyPayment(),
      yearlyPayment: this.getYearlyPayment(),
      totalInterest: 0.0,
      totalPrincipal: 0.0,
      amortizationTableItems: [],
    };

    let beginningBalance = this.amount;
    let totalInterestPaid = 0.0;
    let totalPrincipalPaid = 0.0;
    const yearlyPayment = this.getYearlyPayment();

    for (let i = 0; i < this.terms; i++) {
      const interestPaid = beginningBalance * this.rate;
      totalInterestPaid += interestPaid;
      const principalPaid = yearlyPayment - interestPaid;
      totalPrincipalPaid += principalPaid;
      let remainingPrincipal = beginningBalance - principalPaid;
      remainingPrincipal = remainingPrincipal < 1 ? 0 : remainingPrincipal;

      amortizationTable.amortizationTableItems.push({
        term: i + 1,
        beginningBalance: beginningBalance,
        principalPaid: principalPaid,
        interestPaid: interestPaid,
        remainingPrincipal: remainingPrincipal,
      });

      beginningBalance = remainingPrincipal;
    }

    amortizationTable.totalInterest = totalInterestPaid;
    amortizationTable.totalPrincipal = totalPrincipalPaid;

    return amortizationTable;
  }
}
